File: src/lib/ast/cursor.ts
import { ASTNode, Cursor } from './types'

function getSlot(node: ASTNode, slot: string): ASTNode[] | null {
  const n = node as any
  return (n[slot] as ASTNode[]) ?? null
}

function getList(root: ASTNode[], path: Cursor['path']): ASTNode[] {
  let cur: ASTNode[] = root
  for (const seg of path) {
    cur = getSlot(cur[seg.nodeIndex], seg.slot)!
  }
  return cur
}

function parentList(root: ASTNode[], cursor: Cursor): ASTNode[] {
  return getList(root, cursor.path.slice(0, -1))
}

export function moveCursorRight(root: ASTNode[], cursor: Cursor): Cursor {
  const list = getList(root, cursor.path)

  if (cursor.insertAt < list.length) {
    const next = list[cursor.insertAt]
    if (next.type === 'fraction') {
      return { path: [...cursor.path, { nodeIndex: cursor.insertAt, slot: 'numerator' }], insertAt: 0 }
    }
    if (next.type === 'exponent') {
      return { path: [...cursor.path, { nodeIndex: cursor.insertAt, slot: 'base' }], insertAt: 0 }
    }
    if (next.type === 'radical') {
      return { path: [...cursor.path, { nodeIndex: cursor.insertAt, slot: 'radicand' }], insertAt: 0 }
    }
    if (next.type === 'paren') {
      return { path: [...cursor.path, { nodeIndex: cursor.insertAt, slot: 'children' }], insertAt: 0 }
    }
    return { path: cursor.path, insertAt: cursor.insertAt + 1 }
  }

  if (cursor.path.length === 0) return cursor
  const parentSeg = cursor.path[cursor.path.length - 1]
  const parentPath = cursor.path.slice(0, -1)

  if (parentSeg.slot === 'numerator') {
    return { path: [...parentPath, { nodeIndex: parentSeg.nodeIndex, slot: 'denominator' }], insertAt: 0 }
  }
  if (parentSeg.slot === 'base') {
    return { path: [...parentPath, { nodeIndex: parentSeg.nodeIndex, slot: 'exponent' }], insertAt: 0 }
  }
  return { path: parentPath, insertAt: parentSeg.nodeIndex + 1 }
}

export function moveCursorLeft(root: ASTNode[], cursor: Cursor): Cursor {
  if (cursor.insertAt > 0) {
    const list = getList(root, cursor.path)
    const prev = list[cursor.insertAt - 1]
    if (prev.type === 'fraction') {
      return {
        path: [...cursor.path, { nodeIndex: cursor.insertAt - 1, slot: 'denominator' }],
        insertAt: prev.denominator.length,
      }
    }
    if (prev.type === 'exponent') {
      return {
        path: [...cursor.path, { nodeIndex: cursor.insertAt - 1, slot: 'exponent' }],
        insertAt: prev.exponent.length,
      }
    }
    if (prev.type === 'radical') {
      return {
        path: [...cursor.path, { nodeIndex: cursor.insertAt - 1, slot: 'radicand' }],
        insertAt: prev.radicand.length,
      }
    }
    if (prev.type === 'paren') {
      return {
        path: [...cursor.path, { nodeIndex: cursor.insertAt - 1, slot: 'children' }],
        insertAt: prev.children.length,
      }
    }
    return { path: cursor.path, insertAt: cursor.insertAt - 1 }
  }

  if (cursor.path.length === 0) return cursor
  const parentSeg = cursor.path[cursor.path.length - 1]
  const parentPath = cursor.path.slice(0, -1)

  if (parentSeg.slot === 'denominator') {
    const pList = parentList(root, cursor)
    const frac = pList[parentSeg.nodeIndex] as any
    return {
      path: [...parentPath, { nodeIndex: parentSeg.nodeIndex, slot: 'numerator' }],
      insertAt: frac.numerator.length,
    }
  }
  if (parentSeg.slot === 'exponent') {
    const pList = parentList(root, cursor)
    const exp = pList[parentSeg.nodeIndex] as any
    return {
      path: [...parentPath, { nodeIndex: parentSeg.nodeIndex, slot: 'base' }],
      insertAt: exp.base.length,
    }
  }
  return { path: parentPath, insertAt: parentSeg.nodeIndex }
}

export function moveCursorUp(root: ASTNode[], cursor: Cursor): Cursor {
  if (cursor.path.length === 0) return cursor
  const parentSeg = cursor.path[cursor.path.length - 1]
  const parentPath = cursor.path.slice(0, -1)

  if (parentSeg.slot === 'denominator') {
    const pList = getList(root, parentPath)
    const frac = pList[parentSeg.nodeIndex] as any
    return {
      path: [...parentPath, { nodeIndex: parentSeg.nodeIndex, slot: 'numerator' }],
      insertAt: Math.min(cursor.insertAt, frac.numerator.length),
    }
  }
  if (parentSeg.slot === 'exponent') {
    const pList = getList(root, parentPath)
    const exp = pList[parentSeg.nodeIndex] as any
    return {
      path: [...parentPath, { nodeIndex: parentSeg.nodeIndex, slot: 'base' }],
      insertAt: Math.min(cursor.insertAt, exp.base.length),
    }
  }
  return cursor
}

export function moveCursorDown(root: ASTNode[], cursor: Cursor): Cursor {
  if (cursor.path.length === 0) return cursor
  const parentSeg = cursor.path[cursor.path.length - 1]
  const parentPath = cursor.path.slice(0, -1)

  if (parentSeg.slot === 'numerator') {
    const pList = getList(root, parentPath)
    const frac = pList[parentSeg.nodeIndex] as any
    return {
      path: [...parentPath, { nodeIndex: parentSeg.nodeIndex, slot: 'denominator' }],
      insertAt: Math.min(cursor.insertAt, frac.denominator.length),
    }
  }
  if (parentSeg.slot === 'base') {
    const pList = getList(root, parentPath)
    const exp = pList[parentSeg.nodeIndex] as any
    return {
      path: [...parentPath, { nodeIndex: parentSeg.nodeIndex, slot: 'exponent' }],
      insertAt: Math.min(cursor.insertAt, exp.exponent.length),
    }
  }
  return cursor
}

File: lib/core/ast/cursor.dart
import 'dart:math';
import 'types.dart';

List<ASTNode>? _getSlot(ASTNode node, String slot) {
  switch (node) {
    case FractionNode n:
      if (slot == 'numerator') return n.numerator;
      if (slot == 'denominator') return n.denominator;
    case ExponentNode n:
      if (slot == 'base') return n.base;
      if (slot == 'exponent') return n.exponent;
    case RadicalNode n:
      if (slot == 'degree') return n.degree;
      if (slot == 'radicand') return n.radicand;
    case FunctionNode n:
      if (slot == 'argument') return n.argument;
    case FactorialNode n:
      if (slot == 'operand') return n.operand;
    case NcrNode n:
      if (slot == 'n') return n.n;
      if (slot == 'r') return n.r;
    case NprNode n:
      if (slot == 'n') return n.n;
      if (slot == 'r') return n.r;
    case ParenNode n:
      if (slot == 'children') return n.children;
    default:
      break;
  }
  return null;
}

List<ASTNode> _getList(List<ASTNode> root, List<CursorSegment> path) {
  List<ASTNode> cur = root;
  for (final seg in path) {
    cur = _getSlot(cur[seg.nodeIndex], seg.slot)!;
  }
  return cur;
}

List<ASTNode> _parentList(List<ASTNode> root, Cursor cursor) {
  return _getList(root, cursor.path.sublist(0, cursor.path.length - 1));
}

Cursor moveCursorRight(List<ASTNode> root, Cursor cursor) {
  final list = _getList(root, cursor.path);
  if (cursor.insertAt < list.length) {
    final next = list[cursor.insertAt];
    if (next is FractionNode) {
      return Cursor(
          [...cursor.path, CursorSegment(cursor.insertAt, 'numerator')], 0);
    }
    if (next is ExponentNode) {
      return Cursor(
          [...cursor.path, CursorSegment(cursor.insertAt, 'base')], 0);
    }
    if (next is RadicalNode) {
      return Cursor(
          [...cursor.path, CursorSegment(cursor.insertAt, 'radicand')], 0);
    }
    if (next is ParenNode) {
      return Cursor(
          [...cursor.path, CursorSegment(cursor.insertAt, 'children')], 0);
    }
    return Cursor(cursor.path, cursor.insertAt + 1);
  }
  if (cursor.path.isEmpty) return cursor;
  final parentSeg = cursor.path.last;
  final parentPath = cursor.path.sublist(0, cursor.path.length - 1);
  if (parentSeg.slot == 'numerator') {
    return Cursor(
        [...parentPath, CursorSegment(parentSeg.nodeIndex, 'denominator')], 0);
  }
  if (parentSeg.slot == 'base') {
    return Cursor(
        [...parentPath, CursorSegment(parentSeg.nodeIndex, 'exponent')], 0);
  }
  return Cursor(parentPath, parentSeg.nodeIndex + 1);
}

Cursor moveCursorLeft(List<ASTNode> root, Cursor cursor) {
  if (cursor.insertAt > 0) {
    final list = _getList(root, cursor.path);
    final prev = list[cursor.insertAt - 1];
    if (prev is FractionNode) {
      return Cursor(
          [...cursor.path, CursorSegment(cursor.insertAt - 1, 'denominator')],
          prev.denominator.length);
    }
    if (prev is ExponentNode) {
      return Cursor(
          [...cursor.path, CursorSegment(cursor.insertAt - 1, 'exponent')],
          prev.exponent.length);
    }
    if (prev is RadicalNode) {
      return Cursor(
          [...cursor.path, CursorSegment(cursor.insertAt - 1, 'radicand')],
          prev.radicand.length);
    }
    if (prev is ParenNode) {
      return Cursor(
          [...cursor.path, CursorSegment(cursor.insertAt - 1, 'children')],
          prev.children.length);
    }
    return Cursor(cursor.path, cursor.insertAt - 1);
  }
  if (cursor.path.isEmpty) return cursor;
  final parentSeg = cursor.path.last;
  final parentPath = cursor.path.sublist(0, cursor.path.length - 1);
  if (parentSeg.slot == 'denominator') {
    final frac = _parentList(root, cursor)[parentSeg.nodeIndex] as FractionNode;
    return Cursor(
        [...parentPath, CursorSegment(parentSeg.nodeIndex, 'numerator')],
        frac.numerator.length);
  }
  if (parentSeg.slot == 'exponent') {
    final exp = _parentList(root, cursor)[parentSeg.nodeIndex] as ExponentNode;
    return Cursor(
        [...parentPath, CursorSegment(parentSeg.nodeIndex, 'base')],
        exp.base.length);
  }
  return Cursor(parentPath, parentSeg.nodeIndex);
}

Cursor moveCursorUp(List<ASTNode> root, Cursor cursor) {
  if (cursor.path.isEmpty) return cursor;
  final parentSeg = cursor.path.last;
  final parentPath = cursor.path.sublist(0, cursor.path.length - 1);
  if (parentSeg.slot == 'denominator') {
    final frac = _getList(root, parentPath)[parentSeg.nodeIndex] as FractionNode;
    return Cursor(
        [...parentPath, CursorSegment(parentSeg.nodeIndex, 'numerator')],
        min(cursor.insertAt, frac.numerator.length));
  }
  if (parentSeg.slot == 'exponent') {
    final exp = _getList(root, parentPath)[parentSeg.nodeIndex] as ExponentNode;
    return Cursor(
        [...parentPath, CursorSegment(parentSeg.nodeIndex, 'base')],
        min(cursor.insertAt, exp.base.length));
  }
  return cursor;
}

Cursor moveCursorDown(List<ASTNode> root, Cursor cursor) {
  if (cursor.path.isEmpty) return cursor;
  final parentSeg = cursor.path.last;
  final parentPath = cursor.path.sublist(0, cursor.path.length - 1);
  if (parentSeg.slot == 'numerator') {
    final frac = _getList(root, parentPath)[parentSeg.nodeIndex] as FractionNode;
    return Cursor(
        [...parentPath, CursorSegment(parentSeg.nodeIndex, 'denominator')],
        min(cursor.insertAt, frac.denominator.length));
  }
  if (parentSeg.slot == 'base') {
    final exp = _getList(root, parentPath)[parentSeg.nodeIndex] as ExponentNode;
    return Cursor(
        [...parentPath, CursorSegment(parentSeg.nodeIndex, 'exponent')],
        min(cursor.insertAt, exp.exponent.length));
  }
  return cursor;
}
